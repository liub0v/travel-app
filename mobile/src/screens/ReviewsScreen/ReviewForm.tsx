import React, {useState} from 'react';
import {commentValidationSchema} from '../../services/validation';
import {
  ButtonWrapper,
  CommentInputWrapper,
  CommentTextInput,
  CriterionRatingContainer,
  StarsRatingContainer,
} from './CommentInput.style';
import {CriterionRating} from '../../components/CriterionRating/CriterionRating';
import {StarsRating} from '../../components/StarsRating/StarsRating';
import colors from '../../constants/colors';
import {Text} from 'react-native';
import {ButtonItem} from '../../components/Buttons/ButtonItem';
import {Formik} from 'formik';

export interface FormValues {
  commentText: string;
  interestingRatingValue: number;
  guideRatingValue: number;
  serviceRatingValue: number;
  priceRatingValue: number;
  starsRatingValue: number;
}

interface ProfileFormProps {
  showCriterionRating?: boolean;
  onSubmitHandler: (values: FormValues) => void;
}
export const ReviewForm: React.SFC<ProfileFormProps> = ({
  showCriterionRating,
  onSubmitHandler,
}) => {
  const [starsRatingValue, setStarsRatingValue] = useState(0);

  return (
    <Formik
      validationSchema={commentValidationSchema}
      initialValues={{
        commentText: '',
        interestingRatingValue: 0,
        guideRatingValue: 0,
        serviceRatingValue: 0,
        priceRatingValue: 0,
        starsRatingValue: 0,
      }}
      onSubmit={async values => {
        await onSubmitHandler({...values, starsRatingValue});
      }}>
      {({
        handleChange,
        handleBlur,
        handleSubmit,
        values,
        errors,
        touched,
        setFieldValue,
      }) => (
        <CommentInputWrapper>
          {showCriterionRating && (
            <CriterionRatingContainer>
              <CriterionRating
                testID={'interesting-rating-value'}
                title={'Interesting'}
                ratingValue={values.interestingRatingValue}
                onValueChangeHandler={(value: number) => {
                  setFieldValue('interestingRatingValue', value);
                }}
              />
              <CriterionRating
                testID="guide-rating-value"
                title={'Guide'}
                ratingValue={values.guideRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('guideRatingValue', value)
                }
              />
              <CriterionRating
                testID="service-rating-value"
                title={'Service'}
                ratingValue={values.serviceRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('serviceRatingValue', value)
                }
              />
              <CriterionRating
                testID="price-rating-value"
                title={'Price'}
                ratingValue={values.priceRatingValue}
                onValueChangeHandler={(value: number) =>
                  setFieldValue('priceRatingValue', value)
                }
              />
            </CriterionRatingContainer>
          )}

          <StarsRatingContainer>
            <StarsRating
              setStarRating={setStarsRatingValue}
              initStarsNumber={starsRatingValue}
            />
          </StarsRatingContainer>

          <CommentTextInput
            testID="comment-text-input"
            multiline={true}
            placeholderTextColor={colors.grey}
            placeholder={'Type a comment...'}
            onChangeText={handleChange('commentText')}
            onBlur={handleBlur('commentText')}
            value={values.commentText}
            name={'comment'}
          />
          {errors.commentText && touched.commentText && (
            <Text
              testID="validation-error"
              style={{color: colors.red, padding: 6}}>
              {errors.commentText}
            </Text>
          )}
          <ButtonWrapper>
            <ButtonItem
              title={'Send'}
              titleSize={12}
              size={{height: 40, width: 100}}
              handler={handleSubmit}
            />
          </ButtonWrapper>
        </CommentInputWrapper>
      )}
    </Formik>
  );
};
