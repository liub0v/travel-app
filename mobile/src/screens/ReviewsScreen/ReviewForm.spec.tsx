import React from 'react';
import {ReviewForm} from './ReviewForm';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import {ReactTestInstance} from 'react-test-renderer';

describe('Testing ReviewForm component', () => {
  describe('rendering and submitting a basic Formik form', () => {
    let wrapper: any;
    let onSubmitHandler: () => void;
    let showCriterionRating: boolean;
    onSubmitHandler = jest.fn();
    showCriterionRating = true;
    let commentInput: ReactTestInstance;
    let submitButton: ReactTestInstance;
    const props = {
      onSubmitHandler,
      showCriterionRating,
    };
    beforeEach(() => {
      wrapper = render(<ReviewForm {...props} />);
      commentInput = wrapper.getByTestId('comment-text-input');
      submitButton = wrapper.getByText('Send');
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should changed the text', async () => {
      await waitFor(() => {
        fireEvent.changeText(commentInput, 'test comment');
      });
      fireEvent.press(submitButton);

      await waitFor(() => {
        expect(commentInput.props.value).toEqual('test comment');
      });
    });
    it('should show validation error ', async () => {
      await waitFor(() => {
        fireEvent.changeText(commentInput, ' ');
      });
      fireEvent.press(submitButton);

      const validationErrorText = wrapper.getByTestId('validation-error');
      await waitFor(() => {
        expect(validationErrorText).toBeTruthy();
      });
      expect(validationErrorText.props.children).toBe('Your review is empty');
    });
    it('should not show validation error ', async () => {
      await waitFor(() => {
        fireEvent.changeText(commentInput, 'text');
      });
      fireEvent.press(submitButton);

      await waitFor(() => {
        expect(wrapper.queryByTestId('validation-error')).toBeFalsy();
      });
    });

    it('should send init values if comment is not empty', async () => {
      await waitFor(() => {
        fireEvent.changeText(commentInput, 'text');
      });
      fireEvent.press(submitButton);

      await waitFor(() =>
        expect(onSubmitHandler).toHaveBeenCalledWith({
          commentText: 'text',
          interestingRatingValue: 0,
          guideRatingValue: 0,
          serviceRatingValue: 0,
          priceRatingValue: 0,
          starsRatingValue: 0,
        }),
      );
    });

    it('should send interesting rating ', async () => {
      const interestingRatingSlider = wrapper.getByTestId(
        'interesting-rating-value-slider',
      );
      const guideRatingSlider = wrapper.getByTestId(
        'guide-rating-value-slider',
      );
      const serviceRatingSlider = wrapper.getByTestId(
        'service-rating-value-slider',
      );
      const priceRatingSlider = wrapper.getByTestId(
        'price-rating-value-slider',
      );

      await waitFor(() => {
        fireEvent(interestingRatingSlider, 'onValueChange', 10);
        fireEvent(guideRatingSlider, 'onValueChange', 9);
        fireEvent(serviceRatingSlider, 'onValueChange', 8);
        fireEvent(priceRatingSlider, 'onValueChange', 7);
        fireEvent.changeText(commentInput, 'text');
      });

      fireEvent.press(submitButton);

      await waitFor(() =>
        expect(onSubmitHandler).toHaveBeenCalledWith({
          commentText: 'text',
          interestingRatingValue: 10,
          guideRatingValue: 9,
          serviceRatingValue: 8,
          priceRatingValue: 7,
          starsRatingValue: 0,
        }),
      );
    });
  });
});
