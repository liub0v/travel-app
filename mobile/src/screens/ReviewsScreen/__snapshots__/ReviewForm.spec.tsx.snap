// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ReviewForm component rendering and submitting a basic Formik form should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "paddingLeft": 24,
        "paddingRight": 24,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": 24,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF7F",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        Interesting
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "MontserratExtraBold",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        0
      </Text>
    </View>
    <RNCSlider
      disabled={false}
      enabled={true}
      inverted={false}
      maximumTrackTintColor="#FFFFFF"
      maximumValue={10}
      minimumTrackTintColor="#2D9CDB"
      minimumValue={0}
      onChange={[Function]}
      onRNCSliderSlidingComplete={null}
      onRNCSliderSlidingStart={null}
      onRNCSliderValueChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      step={1}
      style={
        Array [
          Object {
            "height": 40,
          },
          Object {
            "height": 40,
            "width": "100%",
          },
        ]
      }
      tapToSeek={false}
      thumbImage={null}
      thumbTintColor="#2D9CDB"
      value={0}
    />
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF7F",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        Guide
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "MontserratExtraBold",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        0
      </Text>
    </View>
    <RNCSlider
      disabled={false}
      enabled={true}
      inverted={false}
      maximumTrackTintColor="#FFFFFF"
      maximumValue={10}
      minimumTrackTintColor="#2D9CDB"
      minimumValue={0}
      onChange={[Function]}
      onRNCSliderSlidingComplete={null}
      onRNCSliderSlidingStart={null}
      onRNCSliderValueChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      step={1}
      style={
        Array [
          Object {
            "height": 40,
          },
          Object {
            "height": 40,
            "width": "100%",
          },
        ]
      }
      tapToSeek={false}
      thumbImage={null}
      thumbTintColor="#2D9CDB"
      value={0}
    />
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF7F",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        Service
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "MontserratExtraBold",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        0
      </Text>
    </View>
    <RNCSlider
      disabled={false}
      enabled={true}
      inverted={false}
      maximumTrackTintColor="#FFFFFF"
      maximumValue={10}
      minimumTrackTintColor="#2D9CDB"
      minimumValue={0}
      onChange={[Function]}
      onRNCSliderSlidingComplete={null}
      onRNCSliderSlidingStart={null}
      onRNCSliderValueChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      step={1}
      style={
        Array [
          Object {
            "height": 40,
          },
          Object {
            "height": 40,
            "width": "100%",
          },
        ]
      }
      tapToSeek={false}
      thumbImage={null}
      thumbTintColor="#2D9CDB"
      value={0}
    />
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF7F",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        Price
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "MontserratExtraBold",
              "fontSize": 14,
              "paddingLeft": 12,
            },
          ]
        }
      >
        0
      </Text>
    </View>
    <RNCSlider
      disabled={false}
      enabled={true}
      inverted={false}
      maximumTrackTintColor="#FFFFFF"
      maximumValue={10}
      minimumTrackTintColor="#2D9CDB"
      minimumValue={0}
      onChange={[Function]}
      onRNCSliderSlidingComplete={null}
      onRNCSliderSlidingStart={null}
      onRNCSliderValueChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      step={1}
      style={
        Array [
          Object {
            "height": 40,
          },
          Object {
            "height": 40,
            "width": "100%",
          },
        ]
      }
      tapToSeek={false}
      thumbImage={null}
      thumbTintColor="#2D9CDB"
      value={0}
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": 24,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
            "marginBottom": 24,
            "paddingLeft": 24,
            "paddingRight": 24,
          },
        ]
      }
    >
      <Image
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "testUri": "../../../assets/images/star.png",
          }
        }
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
      />
      <Image
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "testUri": "../../../assets/images/star.png",
          }
        }
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
      />
      <Image
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "testUri": "../../../assets/images/star.png",
          }
        }
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
      />
      <Image
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "testUri": "../../../assets/images/star.png",
          }
        }
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
      />
      <Image
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "testUri": "../../../assets/images/star.png",
          }
        }
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
      />
    </View>
  </View>
  <TextInput
    multiline={true}
    name="comment"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Type a comment..."
    placeholderTextColor="#FFFFFF7F"
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 16,
          "color": "#2D9CDB",
          "fontFamily": "Montserrat",
          "fontSize": 14,
          "marginTop": 12,
          "paddingBottom": 12,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 12,
        },
      ]
    }
    testID="comment-text-input"
    value=""
  />
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-end",
          "marginTop": 24,
          "width": "30%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "width": "100%",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2D9CDB",
            "borderRadius": 16,
            "height": 40,
            "justifyContent": "center",
            "opacity": 1,
            "width": "100%",
          }
        }
      >
        <Text
          color="#FFFFFF"
          fontSize={12}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 12,
              },
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</View>
`;
